consensus/spos/consensusMessageValidator.go

```diff
@@ -0,0 +1,270 @@
+		log.Trace("received message from consensus topic has a past round",
+			"msg type", wrk.consensusService.GetStringValue(msgType),
+			"from", cnsMsg.PubKey,
+			"header hash", cnsMsg.BlockHeaderHash,
+			"msg round", cnsMsg.RoundIndex,
+			"round", wrk.consensusState.RoundIndex,
+		)
+
+		return ErrMessageForPastRound
+	}
+
+	err = wrk.checkSignature(cnsMsg)
+	if err != nil {
+		return fmt.Errorf("%w : verify signature for received message from consensus topic failed: %s",
+			ErrInvalidSignature,
+			err.Error())
+	}
+
+	return nil
+}
+
+func (wrk *Worker) checkConsensusMessageValidityForMessageType(cnsMsg *consensus.Message) error {
+	msgType := consensus.MessageType(cnsMsg.MsgType)
+
+	if wrk.consensusService.IsMessageWithBlockBodyAndHeader(msgType) {
+		return wrk.checkMessageWithBlockBodyAndHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockBody(msgType) {
+		return wrk.checkMessageWithBlockBodyValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithBlockHeader(msgType) {
+		return wrk.checkMessageWithBlockHeaderValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithSignature(msgType) {
+		return wrk.checkMessageWithSignatureValidity(cnsMsg)
+	}
+
+	if wrk.consensusService.IsMessageWithFinalInfo(msgType) {
+		return wrk.checkMessageWithFinalInfoValidity(cnsMsg)
+	}
+
+	return fmt.Errorf("%w : received message type from consensus topic is invalid: %d",
+		ErrInvalidMessageType,
+		msgType)
+}
+
+func (wrk *Worker) checkMessageWithBlockBodyAndHeaderValidity(cnsMsg *consensus.Message) error {
+	isMessageInvalid := cnsMsg.SignatureShare != nil ||
+		cnsMsg.PubKeysBitmap != nil ||
+		cnsMsg.AggregateSignature != nil ||
+		cnsMsg.LeaderSignature != nil
+
+	if isMessageInvalid {
+		return fmt.Errorf("%w : received message from consensus topic is invalid: "+
+			"SignatureShare = %s PubKeysBitmap = %s AggregateSignature = %s LeaderSignature = %s",
+			ErrInvalidMessage,
+			logger.DisplayByteSlice(cnsMsg.SignatureShare),
+			logger.DisplayByteSlice(cnsMsg.PubKeysBitmap),
+			logger.DisplayByteSlice(cnsMsg.AggregateSignature),
+			logger.DisplayByteSlice(cnsMsg.LeaderSignature))
+	}
+
+	if len(cnsMsg.Body) > core.MegabyteSize {
+		return fmt.Errorf("%w : received body from consensus topic has an invalid size: %d",
+			ErrInvalidBodySize,
+			len(cnsMsg.Body))
+	}
+
+	if cnsMsg.Header == nil || len(cnsMsg.Header) > core.MegabyteSize {
+		return fmt.Errorf("%w : received header from consensus topic has an invalid size: %d",
+			ErrInvalidHeaderSize,
+			len(cnsMsg.Header))
+	}
+
+	return nil
+}
+
+func (wrk *Worker) checkMessageWithBlockBodyValidity(cnsMsg *consensus.Message) error {
+	isMessageInvalid := cnsMsg.Header != nil ||
+		cnsMsg.SignatureShare != nil ||
+		cnsMsg.PubKeysBitmap != nil ||
+		cnsMsg.AggregateSignature != nil ||
+		cnsMsg.LeaderSignature != nil
+
+	if isMessageInvalid {
+		return fmt.Errorf("%w : received message from consensus topic is invalid: "+
```

Comment:
is this error message not too long ? maybe do a log trace before then send and simpler wrapped error.

Reply:
Done